text/0161-pton-standard.md

```diff
@@ -0,0 +1,287 @@
+- **TEP**: [161](https://github.com/ton-blockchain/TEPs/pull/161)
+- **title**: Proxy TON for DeFi applications
+- **status**: Draft
+- **type**: Contract Interface
+- **authors**: [STON.fi](https://ston.fi), [Dario Tarantini](https://github.com/dariotarantini)
+- **created**: 13.06.2024
+- **replaces**: -
+- **replaced by**: -
+
+# Summary
+
+A standard interface for tokenized TON. Partially implement TEP-74 standard.
+
+# Motivation
+
+A standard interface will greatly simplify interaction and usage of TON in dApps. An application can only focus on handling standard jetton calls, while still being able to use native TON without the need of additional logic for it.
+
+# Guide
+
+## Useful links
+
+1. [Reference pTON implementation](https://github.com/ston-fi/pton-contracts)
+
+# Specification
+
+Proxy TON must be organized as follows: there is minter smart-contract which is used to mint new wallets and provide common information. Total supply must be 0, since minter can't know how many tokenized tokens there are in all wallets.
+
+At the same time information about amount of tokenized TON owned by each user is stores in decentralized manner in individual (for each owner) smart-contracts called "wallets".
+
+## Wallet Contract
+
+Must implement:
+
+### Internal message handlers
+
+#### 1. `ton_transfer`
+
+##### Request
+
+TL-B schema of inbound message:
+
+```
+ton_transfer#1f3835d
+   query_id:uint64 
+   ton_amount:Coins 
+   refund_address:MsgAddress 
+   forward_payload:(Either Cell ^Cell) 
+= InternalMsgBody;
+```
+
+`query_id` - arbitrary request number.
+
+`ton_amount` - amount of transferred TON in elementary units.
+
+`refund_address` - address of the the refund tx, if execution fails.
+
+`forward_payload` - optional custom data that should be sent to the wallet owner address.
+
+**Should be rejected if:**
+
+1. `ton_amount` is zero
+2. `refund_address` is not `addr_std`
+3. message value is less or equal than `ton_amount`
+4. outgoing messages fails
+5. if rejected, should send a `ton_refund` message with following layout TL-B: `ton_refund#ae25d79e query_id:uint64 = InternalMsgBody`
+
+**Otherwise should do:**
+
+1. increase internal jetton balance by `ton_amount`
+2. the receiver's jetton-wallet send message to `owner` address with unused gas attached and with the TEP-74 `transfer_notification` message body
```

Comment:
it’s not very clear what `amount` should be in transfer_notification. equal to `ton_amount`?